LiteCore/Database/DatabaseImpl+Upgrade.cc

```diff
@@ -54,6 +54,31 @@ namespace litecore {
     }
 
 
+namespace {
+    struct RestoreIndex {
+        DatabaseImpl* _db;
+        ExclusiveTransaction& _transaction;
+        vector<IndexSpec> _specs;
+
+        RestoreIndex(DatabaseImpl* db, ExclusiveTransaction& t)
+            : _db(db)
+            , _transaction(t)
+            , _specs{_db->defaultKeyStore().getIndexes()}
+        {
+            for (IndexSpec& spec: _specs) {
+                _db->defaultKeyStore().deleteIndex(spec.name, _transaction);
+            }
+        }
+
+        ~RestoreIndex() {
```

Comment:
We should avoid putting anything in a destructor that might throw an exception, since if that destructor is called as cleanup during exception handling ,and it throws another exception, C++ will terminate the process.

There isn't any need to restore the indexes on an exception, because the caller will just abort the transaction anyway.

You can just put the delete-indexes and restore-indexes code directly into upgradeDocumentVersioning.

Reply:
Good point. I will do as you suggesting.